### ISSUE TITLE
Add a solution for large select fields in webforms

### ISSUE THREAD

**Issue by rooby** [govCMS/GovCMS7#284]
If you have a form with a lot of options (in my current use case, it is almost 3000 in a single field) then a normal select field is unwieldy to manage. It also means you have to have the webform componenet permission to edit the values, whereas with taxonomy your normal content editors can manage the options.

You also run into the size limit of the {webform_component.extra} database column that stores the data, meaning you then have to split it into multiple fields.

It is also not great for a user to have to navigate a select field with thousands of items.

There are a couple of potential solutions to this:
- [Webform term options](https://www.drupal.org/project/webform_term_opts): Adds the ability to use a vocabulary as a predefined option list for a select field.
- [Webform autocomplete](https://www.drupal.org/project/webform_autocomplete): Adds autocomplete text fields, also with the ability to [use a vocabulary as the source](https://www.drupal.org/node/2624466).

I think it might be useful to include both modules because if you just want to maintain taxonomy select fields the webform term options module would be useful but if you also had very large datasets then the autocomplete is needed.

(Issue #147 is related to this)

**Comment by nathan-w** [comment]
Its actually not great for users to interact with just about any select box. You definitely shouldn't use them when you've got 3000 items to pick from, but you also shouldn't use them if you only have 2 or 3 things in a list. A short list of things could just be radio buttons with large click targets.

Drop-down boxes, by their very nature, hide information and it must be unpacked and sorted through by the user in order to find the correct response. Depending on the device you're using interacting with a select box can mean multiple interactions/clicks/touches. You also need a high degree of precision which isn't so good with people who have physical conditions like poor motor control - or in my case - BIG fingers.

Where there are many possible options e.g. in a list of countries, you should definitely implement a search feature that allows users to filter the options. @rooby Im not sure the WTO solves this problem, especially if the field is still tied to a large vocabulary. 

I've used autocomplete before and it sort of works, you still have to have some idea of what term you're looking for, and how to spell it. 

There might be some other solutions around. It's worth looking at IMO.

**Comment by rooby** [comment]
Definitely a select box is not good for my 3000 option field. Autocomplete is much better, especially seeing as in this particular context the end user will know what they are looking for in that list.

In this same webform though there are also smaller lists of between 10 and 40 items, for which a select box may be appropriate but from a content administrator point of view vocabularies is a much better way to manage those lists, which may also appear elsewhere on the site.

If you don't have any idea what you're looking for in a list of thousands then you're probably not going to have a great time with any of the options.

**Comment by nathan-w** [comment]
@rooby are you using the [Chosen](https://www.drupal.org/project/chosen) module? That's included in the base govCMS distro. Won't solve all the usability problems with select fields, but looks like a nice slick autocomplete filter.

**Comment by fiasco** [comment]
Because the webform module is rather bespoke (It was born Drupal 6, before entities and the birth of entityform module which supports fieldable forms), it doesn't support Chosen module 👎 . Instead, it requires its own bespoke suite of modules to support it.

Ideally, entityform module would be the best fit but its not in govCMS. And since we already have a form building module, its not likely we'll add it as two form building modules is technical debt and the upgrade path to replace webform with entityform is a fairly large piece of work I have no confidence we could actually achieve without regression.

So options I see here are:
- Use Webform Autocomplete
- Build your own UX widget in the theme to sit on top of a 3000 choice select list
- Do nothing (always an option)
- Use [Webform Views Select](https://www.drupal.org/project/webform_views_select) which lets the select list be drive from a data source defined by a View (doesn't do autocomplete, but could be contributed)

I think Webform Autocomplete will have a hardtime getting into govCMS because of its functionality that can expose past submission data as autocomplete options. This configuration would be seen as  a high risk data leak. We would need to mitigate it in the PR to ensure it wasn't even a possibility in the distribution.

**Comment by nathan-w** [comment]
@fiasco looking forward to lots more conversations like this :) be good to get an estimate of effort for either adding to views select - or forking and mitigating autocomplete

**Comment by rooby** [comment]
Yeah chosen is nice, I am using it on some fields. In theory you should be able to configure chosen to apply to the webform select elements but it seems to be buggy and doesn't work currently.

@fiasco 
- Use Webform Autocomplete
  - I see your point with that past submissions feature. Autocomplete is the less critical part of this request anyway, so maybe this one is just off the table for now.
- Build your own UX widget in the theme to sit on top of a 3000 choice select list
  - I don't think this is possible with GovCMS. Not via the theme anyway, only if a custom module could get it (getting the terms as options).
- Do nothing (always an option)
  - Not a great option though
- Use Webform Views Select which lets the select list be drive from a data source defined by a View (doesn't do autocomplete, but could be contributed)
  - This is a possibility, and in combination with chosen (when they iron out their webform issues) would suffice.

There is also the 5th option of the Webform term options module.
That module combined with chosen would also be fine although not as flexible as webform views select.

Just to clarify, the main purpose of this is to make webform selects work at all with large option sets. Any enhancement of the UI is just a nice to have.
